src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/MagicLinkLoginFactory.php

```diff
@@ -0,0 +1,127 @@
+                ->info('The lifetime of the magic link in seconds')
+            ->end()
+            ->scalarNode('token_parameter')
+                ->defaultValue('token')
+            ->end()
+            ->scalarNode('success_handler')
+                ->info(sprintf('A service id that implements %s', AuthenticationSuccessHandlerInterface::class))
+            ->end()
+            ->scalarNode('failure_handler')
+                ->info(sprintf('A service id that implements %s', AuthenticationFailureHandlerInterface::class))
+            ->end()
+        ;
+
+        foreach (array_merge($this->defaultSuccessHandlerOptions, $this->defaultFailureHandlerOptions) as $name => $default) {
+            if (\is_bool($default)) {
+                $builder->booleanNode($name)->defaultValue($default);
+            } else {
+                $builder->scalarNode($name)->defaultValue($default);
+            }
+        }
+    }
+
+    public function getKey()
+    {
+        return 'magic-link';
+    }
+
+    public function createAuthenticator(ContainerBuilder $container, string $firewallName, array $config, string $userProviderId): string
+    {
+        if (!$container->hasDefinition('security.authenticator.magic_link')) {
+            $loader = new PhpFileLoader($container, new FileLocator(\dirname(__DIR__).'/../../Resources/config'));
+            $loader->load('security_authenticator_magic_link.php');
+        }
+
+        $linkerId = 'security.authenticator.magic_login_linker.'.$firewallName;
+        $container
+            ->setDefinition($linkerId, new ChildDefinition('security.authenticator.abstract_magic_login_linker'))
+            ->replaceArgument(0, new Reference($config['storage_service']))
+            ->replaceArgument(3, $config['lifetime'])
+            ->replaceArgument(4, $config['check_route'])
+            ->addTag('security.authenticator.magic_linker', ['firewall' => $firewallName])
+        ;
+
+        $authenticatorId = 'security.authenticator.magic_link.'.$firewallName;
+        $container
+            ->setDefinition($authenticatorId, new ChildDefinition('security.authenticator.magic_link'))
+            ->replaceArgument(0, new Reference($linkerId))
+            ->replaceArgument(2, new Reference($this->createAuthenticationSuccessHandler($container, $firewallName, $config)))
+            ->replaceArgument(3, new Reference($this->createAuthenticationFailureHandler($container, $firewallName, $config)))
+            ->replaceArgument(4, [
+                'check_route' => $config['check_route'],
+                'token_parameter' => $config['token_parameter'],
+            ]);
+
+        return $authenticatorId;
+    }
+
+    public function getPosition()
+    {
+        return 'form';
+    }
+
+    protected function createAuthProvider(ContainerBuilder $container, string $id, array $config, string $userProviderId)
+    {
+        throw new \Exception('The old authentication system is not supported with magic_link.');
+    }
```

Comment:
In this PR, I realized that the base class - `AbstractFactory` (which is tied to the OLD authentication system) needs to be "split" a little bit - there are a few places where we still require this class, even if you don't want to implement the old system (e.g. the security `Configuration` class calls the `AbstractFactory::addConfiguration()` method). Anyways, that's for another PR - I shouldn't need to extend `AbstractFactory`, but I still do until we clean that up.

Reply:
Do we really need to extends AbstractFactory here? Can't we "just" implement the interfaces?

Reply:
this may be more possible than I thought - it's on the todo list